### ISSUE TITLE
Warn about redundant `const` in `constexpr` declarations?

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#83837]
Consider:

```
constexpr int* const p = ...;
```

The `const` is redundant because `constexpr` implies constness.

We have some code like this where it seems the author may have intended `p` to be a "constant pointer to a constexpr int", which doesn't really make sense.

Perhaps a warning explaining that `p` is already const because of the constexpr would be helpful both in removing the redundancy, and showing users that constexpr applies to the variable?

**Comment by Atousa** [comment]
I think that second const means that p is a constant pointer that points to an constant integer. It ensures that both the pointer itself and the value it points to are not modified after initialization.

**Comment by zmodem** [comment]
> I think the second 'const' means that 'p' is also a constant integer. It ensures that both the pointer itself and the value it points to are not modified after initialization.

No, both `constexpr` and `const` above refer to the `p` variable itself, not the data it points to. For example, the following code is valid:

```
int i;
constexpr int* const p = &i;

void f() {
  *p = 42;
}
```

This is why I think a warning about the redundant `const` would be helpful :-)

**Comment by AaronBallman** [comment]
I think such a diagnostic would be reasonable to add, but it might be more appropriate for clang-tidy as it seems like a somewhat style-oriented diagnostic.

In C++: https://eel.is/c++draft/dcl.constexpr#6.sentence-1
In C23 6.7.2p16: "An object declared with a storage-class specifier constexpr has its value permanently fixed at translation-time; if not yet present, a const-qualification is implicitly added to the object’s type."

So the `const` specifier is redundant per specification, but it's also harmless in terms of code semantics.

We currently diagnose redundant `const` qualification when you duplicate the qualifier: https://godbolt.org/z/8YKao1s3a so there's precedent for us warning on redundant constructs, but in that case, the diagnostic is old (Clang 3.2, which kind of predates us having much of a policy about diagnostics), is diagnosed by default in GCC (so there's a compat story there), and the extra qualifier usually sneaks in via a macro expansion rather than intentionally.

In this case, it seems reasonable for a user to prefer `constexpr int * const` over `constexpr int *` because the code is more readable to non-language lawyers who don't know that `constexpr` implies `const` qualification, but it also seems reasonable to allow people to catch redundant constructs.

CC @carlosgalvezp @njames93 @PiotrZSL @LegalizeAdulthood for opinions on whether this might make sense for clang-tidy.

**Comment by PiotrZSL** [comment]
We got check for redundant inline, there is nothing bad in having one for redundant const. That's easy check anyway.

**Comment by zmodem** [comment]
I think the real value I was looking for is catching cases where the developer wrote

```
constexpr int* const p = ....;
```

but what they really wanted was to make both p and the pointee const:

```
constexpr const int* p = ...;
```

Even though constexpr is all the way to the left, it's still just applying to p.

That is, it's not to nitpick about applying `const` twice, it's about drawing attention to the possibly unintended semantics. Maybe we could do something like:

warning: this const qualifier on p is redundant
note: because this constexpr already made it const
hint: if you wanted to make the pointee const, move the qualifier here

But if people often prefer to spell out the "implicit" const, then perhaps this would be too noisy. I guess we'd have to prototype something and try it out.

**Comment by carlosgalvezp** [comment]
Such a check makes sense to me, could be in the `readability` section.
